spec/models/transformation_mapping_spec.rb

```diff
@@ -1,37 +1,66 @@
 RSpec.describe TransformationMapping, :v2v do
-  let(:src_ems) { FactoryBot.create(:ems_vmware) }
-  let(:dst_ems_redhat) { FactoryBot.create(:ems_redhat) }
+  let(:src_ems_vmware)    { FactoryBot.create(:ems_vmware) }
+  let(:dst_ems_redhat)    { FactoryBot.create(:ems_redhat) }
   let(:dst_ems_openstack) { FactoryBot.create(:ems_openstack) }
-  let(:src_cluster) { FactoryBot.create(:ems_cluster, :ext_management_system => src_ems) }
+
+  let(:src_cluster_vmware) { FactoryBot.create(:ems_cluster, :ext_management_system => src_ems_vmware) }
   let(:dst_cluster_redhat) { FactoryBot.create(:ems_cluster, :ext_management_system => dst_ems_redhat) }
+
+  let(:src_hosts_vmware) { FactoryBot.create_list(:host_vmware, 1, :ems_cluster => src_cluster_vmware) }
+  let(:dst_hosts_redhat) { FactoryBot.create_list(:host_redhat, 1, :ems_cluster => dst_cluster_redhat) }
+
+  let(:src_storages_vmware) { FactoryBot.create_list(:storage, 1, :hosts => src_hosts_vmware) }
+  let(:dst_storages_redhat) { FactoryBot.create_list(:storage, 1, :hosts => dst_hosts_redhat) }
+
+  let(:src_lan_vmware) { FactoryBot.create(:lan) }
+  let(:dst_lan_redhat) { FactoryBot.create(:lan) }
+
   let(:dst_cloud_tenant_openstack) { FactoryBot.create(:cloud_tenant, :ext_management_system => dst_ems_openstack) }
-  let(:vm) { FactoryBot.create(:vm_vmware, :ems_cluster => src_cluster) }
 
-  let(:mapping_redhat) do
+  let(:mapping_redhat) { FactoryBot.create(:transformation_mapping) }
+  let(:mapping_openstack) { FactoryBot.create(:transformation_mapping) }
+
+  before do
     FactoryBot.create(
-      :transformation_mapping,
-      :transformation_mapping_items => [TransformationMappingItem.new(:source => src_cluster, :destination => dst_cluster_redhat)]
+      :transformation_mapping_item,
+      :source                 => src_cluster_vmware,
+      :destination            => dst_cluster_redhat,
+      :transformation_mapping => mapping_redhat
+    )
```

Comment:
Why switch this from being defined with the mapping to a before each?  Doing `FactoryBot.create(:transformation_mapping, :transformation_mapping_items => [FactoryBot.create(:transformation_mapping_item, ...)])` is much better, this looks like it'll create a new mapping_item everytime a test runs.

Reply:
Originally that's what I tried. The issue is an (upcoming) validation on the `TransformationMappingItem` model which triggers before the association between the mapping and the mapping items is established. That's why I created the mapping first, and then the mapping items in the `before` block.

I tried messing around with `build` vs `create` and adding an association in the factory itself, but nothing really worked.

Reply:
So you need the mapping first, then the mapping items?  Can you just use `.tap {}` then?  This isn't the right way to fix this because it'll create these before every test runs probably causing other issues.